### ISSUE TITLE
Proposal: allow enforced defaults from the Windows registry

### ISSUE THREAD

**Issue by jsachs** [keepassxreboot/keepassxc#2189]
Currently, it is difficult to enforce organizational defaults on Windows. This proposal would implement behavior where, if a certain value is set in the Windows registry, KeePassXC will use the registry instead of the ini config file. Otherwise, it will fall back to using IniFormat.

## Current Behavior

All settings use IniFormat. This makes it difficult to enforce organizational defaults for Windows users of KeePassXC.

## Possible Solution

KeePassXC would check that a value in the registry (HKEY_CURRENT_USER\Software\KeePassXC\OrganizationDefaults) exists. If it does, [m_setttings](https://github.com/keepassxreboot/keepassxc/blob/develop/src/core/Config.cpp#L120) will be initialized with NativeFormat, rather than IniFormat.

## Context

The main use case for this feature is enforcing an idle timeout for all of our users. We currently have the ability to change the default behavior, but not to enforce the configuration option, which we could do using the Windows Registry.

Operating system: Windows 7 - Windows 10

**Comment by droidmonkey** [comment]
Great idea, although it would be far easier to have a registry key that forces a particular ini file (network share/default profile) so no other config changes need to be made in code. Would this work for you?

**Comment by jsachs** [comment]
@droidmonkey how would your scenario work if we only want to enforce certain options, such as locking timeout? The advantage to using the registry seems to be that we can be selective about properties that are enforced.

**Comment by thenktor** [comment]
Is anything like this possible in the meantime? I'd like to disable update check for the whole organization.

**Comment by droidmonkey** [comment]
You'd have to control their entire config file which locks out many other settings. So no, this is not possible yet.

**Comment by thenktor** [comment]
OK, then the only solution so far is a script that runs on user login and changes the values in `%APPDATA%\KeePassXC\keepassxc.ini`.
Thanks anyway for the info!

**Comment by QMUL-JayHay** [comment]
Hi, all. I noticed that the wording implies that this is not an issue / is available for Linux and Mac, but I couldn't find anything on tiered configuration setups for those either. Have I misunderstood the scope here?

**Comment by Saxomania** [comment]
Hi, is there any progress at this topic @droidmonkey? We would like to deploy KeepassXC in our company software store but a not forced configuration is a real disadvantage compared to Keepass. And scripting the user ini file is a bad workaround unless you cant really force and lock settings.

**Comment by droidmonkey** [comment]
None made at this time.

**Comment by Saxomania** [comment]
do you have any timeline? What do you need to get it done?

**Comment by droidmonkey** [comment]
If you have a cloning machine, that would be very useful ☺️

**Comment by Saxomania** [comment]
> If you have a cloning machine, that would be very useful ☺️

☺️

**Comment by htcfreek** [comment]
Two things on this:

1. Everything saved in the Registry under HKCU or the user folder (%Appdata%) can be changed by the user without having administrative permissions. It is not enforced.
2. To enforce settings they have to be saved at a place where normal users don't have write permissions.
3. I suggest to have two files in the application folder to solve this and other "global config" problems. For best flexibility each entry should be evaluated individually.
   -  `keepassxc_defaults.ini` (not enforced but used for settings initialisation)
   - `keepassxc_enforced.ini` (enforced and override all other config files.)

**Comment by droidmonkey** [comment]
HKLM is where we would look for these settings. Adding config files to an installation folder is not going to happen. At worst we would consider ProgramData subfolder. For macOS and Linux, I am open to suggestions.

**Comment by thenktor** [comment]
As macOS and Linux also use the `keepassxc.ini` I actually would not use the registry on Windows.

Usining a `keepassxc_enforced.ini` is a simple way to overwrite settings of a `keepassxc.ini`. An admin just has to set correct file permission rights on that file.

**Comment by htcfreek** [comment]
> As macOS and Linux also use the `keepassxc.ini` I actually would not use the registry on Windows.
>
> Usining a `keepassxc_enforced.ini` is a simple way to overwrite settings of a `keepassxc.ini`.

I agree.

> An admin just has to set correct file permission rights on that file.

This happens automatically when `C:\ProgramData` or the installation directory is used. I thought for compatibility with macOS and Linux the program folder is the easiest way to go.

**Comment by htcfreek** [comment]
> HKLM is where we would look for these settings. Adding config files to an installation folder is not going to happen.

@droidmonkey 
Why? What is the reason for not using a file.

**Comment by QMUL-JayHay** [comment]
> As macOS and Linux also use the `keepassxc.ini` I actually would not use the registry on Windows.
> 
> Usining a `keepassxc_enforced.ini` is a simple way to overwrite settings of a `keepassxc.ini`. An admin just has to set correct file permission rights on that file.

From an IT deployment perspective, expanding the ini model with hierarchical precedence would be an elegant solution for us in deploying KeepassXC across Windows, Mac & Linux. 

Deploying defaults and enforced settings separately (as per the above _defaults.ini) is a good idea too. I did see that you can run Keepass with the -o option to load a config. Does this persist after first run? If so, perhaps there's a way to initialise this way. Perhaps another option is to set a "Default Roaming Config Template" environment variable with the default template configuration as well?

**Comment by droidmonkey** [comment]
> > HKLM is where we would look for these settings. Adding config files to an installation folder is not going to happen.
> 
> @droidmonkey 
> Why? What is the reason for not using a file.

Almost all policy enforcement on Windows is done via the registry. That goes for GPO and Intune management. It's just the way Windows "prefers" to handle security policies. 

Any solution we implement would likely be both file based and registry based.

**Comment by Saxomania** [comment]
any solution is welcome and yes i would prefer also registry over file but then you should use hklm\software\policies instead of hklm\software so even administrators are not able to change settings directly on the client (which is the weak point if you do it via file like keepass is doing it)

**Comment by QMUL-JayHay** [comment]
The registry route will likely take considerably more time to implement and only covers one OS, so I'd imagine the file method would be the way to go first with a view to enhancing Windows specific functionality further down the line. I see there is appetite for this, given the incorporation of Windows Hello for authentication.

**Comment by thenktor** [comment]
> Almost all policy enforcement on Windows is done via the registry. That goes for GPO and Intune management. It's just the way Windows "prefers" to handle security policies.

As someone who has to manage a bunch of Windows computers with software from various vendors I can say that it is a mess anyways :smile:
Rule of thumb: The more expensive the software is, the bigger is the mess managing it :see_no_evil:

**Comment by htcfreek** [comment]
> i would prefer also registry over file but then you should use hklm\\software\\policies instead of hklm\\software so even administrators are not able to change settings directly on the client

Then admx files have to be provided. Because this policy path is not write protected for administrative users. The important point ist that gpos are reapplied periodically.

**Comment by Saxomania** [comment]
> Then admx files have to be provided. Because this policy path is not write protected for administrative users. The important point ist that gpos are reapplied periodically.

even with admin rights you cant change policies path. You have to be SYSTEM user to do so. Yes, if you have admin rights you also can get SYSTEM rights but you need to know how and most admins dont know so its an extra step to achieve this.

And, yes admx files would be the icing on the cake.

**Comment by Saxomania** [comment]
@droidmonkey hi, any news here? Maybe some ETA? Yes i am still full of hope :)

**Comment by droidmonkey** [comment]
Haven't worked on this at all, sorry.

**Comment by Saxomania** [comment]
hope is gone

**Comment by toddalbers** [comment]
What about saving override settings in each KeePassXC database in an additionally encrypted field * along with a check\hash?  
'* Calling it: Encrypted Database Settings Field (EDSF)'

WHEN USER OPENS THE DATABASE:
When the user open the application and\or opens another KeePassXC database...
Then the application does the following:
1. unencrypts this additionally encrypted EDSF field (which are the override settings and the check\hash).  See "Settings Encryption Key\Password" below.
2. Checks the settings against the check\hash (maybe not needed.  Or maybe better stored encrypted elsewhere in the file.)
3. If all checks out, it sets all the values that are in the EDSF and greys out and makes them read-only to the user.

Settings Encryption Key\Password
The Settings Encryption Key\Password would be a special field of which the values can only be changed by an "Admin" account that logs into the database.  So, there would need to be a new "Admin" account per database of which the "KeePassXC Database Admin" account and it's password (changeable by that account) would be stored in the database as a hash (with salt) of the password.  The Encryption Key\Password settings value could possibly be an additional hash -with salt (or a FEW additional hashes-with salt) of the hash of the "Admin" password for that database.

So, 
- your settings in KeePassXC would be dependent upon the database that you have open.
- And (depending on which particular KeePassXC database you have open) you may not be able to change some settings for that database.
- And administrators could pre-create a KeePassXC database using that databases "admin" account (and authentication) and then  publish that database either as the default database of KeePassXC or as a shared KeePassXC  database. 

This obviously is some significant rework within the application itself.  But, if it were all stored in each KeePassXC database, it would be platform agnostic (including registry issues etc.) and I believe meet the need here.

It's just an idea that I am offering here.  But, I "think" this might work?  Of course this doesn't solve the problem that it still requires man-hours of work on the application.  But, maybe this is more "do-able"?

Or maybe it's not do-able at all.  I don't know what I don't know.  Thoughts?  Feedback?

**Comment by droidmonkey** [comment]
Saving settings internal to the database is a different request. This is about system administrators setting defaults for the application itself.

**Comment by Saxomania** [comment]
hi, now that passkeys is added 👍 maybe there is a time to implement a forced configuration? 🏃‍♂️

**Comment by Carlos300000** [comment]
> HKLM is where we would look for these settings. Adding config files to an installation folder is not going to happen. At worst we would consider ProgramData subfolder. For macOS and Linux, I am open to suggestions.

ProgramData is not a good idea, because users have write access there, too. Maybe have a look at how the old Keepass does it. The two suggested files (_default and _enforced) sound perfect for the job.

As this is only relevant for larger organizations, are you open to pull requests to make this possible? I'm not a developer, but I'm sure that there will be someone out there in the postition to help.

**Comment by perfektionizt** [comment]
> > HKLM is where we would look for these settings. Adding config files to an installation folder is not going to happen. At worst we would consider ProgramData subfolder. For macOS and Linux, I am open to suggestions.
> 
> ProgramData is not a good idea, because users have write access there, too. Maybe have a look at how the old Keepass does it. The two suggested files (_default and _enforced) sound perfect for the job.
> 
> As this is only relevant for larger organizations, are you open to pull requests to make this possible? I'm not a developer, but I'm sure that there will be someone out there in the postition to help.

That's not right. “Normal” users do not have write access to %ProgramData%; only local administrators have this access.

For business relevance, either the registry, %ProgramData%, %ProgramFiles%, or %ProgramFiles(x86)% are always used for package/software deployments, and it is common for the logged-in user NOT to have elevated rights (local admin).

**Comment by Saxomania** [comment]
> That's not right. “Normal” users do not have write access to %ProgramData%; only local administrators have this access.

even if you deny it but its still true - as default users have write rights in %ProgramData%
 
> For business relevance, either the registry, %ProgramData%, %ProgramFiles%, or %ProgramFiles(x86)% are always used for package/software deployments, and it is common for the logged-in user NOT to have elevated rights (local admin).

how do you want to deploy software to the registry :) - you should never install software to %ProgramData%

**Comment by QMUL-JayHay** [comment]
Hi, all.

I still can't see from the above why everyone has a problem with an (e.g.) "_override-xxx.ini" file with only the settings you want to override in program files. I've seen applications use file location as a precendence and ini file can just contain the subset of settings that you want to override:
Highest Priority - Program Installation Directory
Lowest Priority - User Appdata

You could also have the default/global standard ini and an optional settingsoverride.ini...
... or naming convention for modular settings enforcement with multiple override ini's : e.g. "_override-confidential.ini", "_override-topsecret.ini" . That said, if the ini was parseable in a standard format, like json/xml etc. you could just use Powershell etc. and manage configuration management that way.

I understand balancing dev resource against enterprise features is important. It would be good to get some commercial donations going or something that fits for you, if that gets this implemented.

That said, keep up the good work. KeepassXC is tight!

**Comment by perfektionizt** [comment]
> > That's not right. “Normal” users do not have write access to %ProgramData%; only local administrators have this access.
> 
> even if you deny it but its still true - as default users have write rights in %ProgramData%
> 
If you need to actively deny local admin rights, then you are doing it wrong. Default users are users without elevated rights and therefore, they don't have access there. "Your" (default) business users should not have admin rights.

> > For business relevance, either the registry, %ProgramData%, %ProgramFiles%, or %ProgramFiles(x86)% are always used for package/software deployments, and it is common for the logged-in user NOT to have elevated rights (local admin).
> 
> how do you want to deploy software to the registry :) - you should never install software to %ProgramData%

It is not a matter of installing the software in the registry, but rather of setting keys in the registry that refer to forced settings for the app. Your argument is based on your own perspective as a private/personal logged-in user who – apparently – has administrator rights as a “standard user.” When rolling out software (installations), the user cannot choose where the software is installed because the default installation path is determined by the programmer.

“Never install software in %ProgramData%” => that is in some cases incorrect and is simply decided by code.

**Comment by Saxomania** [comment]
you denied that users have write rights in %programdata% but you are wrong and thats it. Get a new Windows installation and login as default user and create a new folder in %programdata% - it will work. Guess why default applocker rules allow executing only in programfiles, programfiles(x86) and %windows%. But we should not discuss this here.

**Comment by perfektionizt** [comment]
> you denied that users have write rights in %programdata% but you are wrong and thats it. Get a new Windows installation and login as default user and create a new folder in %programdata% - it will work. Guess why default applocker rules allow executing only in programfiles, programfiles(x86) and %windows%. But we should not discuss this here.

You are still argumenting from the POV of a home Windows user. Like i said: from the business POV, that is not the case.

**Comment by Saxomania** [comment]
yes thats why i want to have a forced config file because this is necessary as a Windows home user. Wrong again. So wrong. Buts thats for me now. I will not answer this anymore. Basta.

**Comment by thenktor** [comment]
> That's not right. “Normal” users do not have write access to %ProgramData%; only local administrators have this access.

A standard user has write access to %ProgramData% (sorry, German output):
```txt
PS C:\> icacls.exe .\ProgramData\
.\ProgramData\ NT-AUTORITÄT\SYSTEM:(OI)(CI)(F)
               VORDEFINIERT\Administratoren:(OI)(CI)(F)
               ERSTELLER-BESITZER:(OI)(CI)(IO)(F)
               VORDEFINIERT\Benutzer:(OI)(CI)(RX)
               VORDEFINIERT\Benutzer:(CI)(WD,AD,WEA,WA)
```
The last line allows standard users to write/append to subfolders/files.

But even Microsoft happily sets other permissions to subfolders during installation of programs, e.g. OpenSSH:
```txt
PS C:\> icacls.exe .\ProgramData\ssh\
.\ProgramData\ssh\ NT-AUTORITÄT\SYSTEM:(OI)(CI)(F)
                   VORDEFINIERT\Administratoren:(OI)(CI)(F)
                   NT-AUTORITÄT\Authentifizierte Benutzer:(OI)(CI)(RX)
```

**Comment by perfektionizt** [comment]
> > That's not right. “Normal” users do not have write access to %ProgramData%; only local administrators have this access.
> 
> A standard user has write access to %ProgramData% (sorry, German output):
> 
> PS C:\> icacls.exe .\ProgramData\
> .\ProgramData\ NT-AUTORITÄT\SYSTEM:(OI)(CI)(F)
>                VORDEFINIERT\Administratoren:(OI)(CI)(F)
>                ERSTELLER-BESITZER:(OI)(CI)(IO)(F)
>                VORDEFINIERT\Benutzer:(OI)(CI)(RX)
>                VORDEFINIERT\Benutzer:(CI)(WD,AD,WEA,WA)
> The last line allows standard users to write/append to subfolders/files.
> 
> But even Microsoft happily sets other permissions to subfolders during installation of programs, e.g. OpenSSH:
> 
> PS C:\> icacls.exe .\ProgramData\ssh\
> .\ProgramData\ssh\ NT-AUTORITÄT\SYSTEM:(OI)(CI)(F)
>                    VORDEFINIERT\Administratoren:(OI)(CI)(F)
>                    NT-AUTORITÄT\Authentifizierte Benutzer:(OI)(CI)(RX)

Again, i‘m not talking about private scenarios. I am aware of the fact, that a regular home user will use his account with admin rights (by following standard Win OS Install procedure).

Maybe the requestor of this topic can chime in and clarify his described context more clearly.

To be fair:
I was mostly wrong. In the context of the logged-in user, write access is available for HIS folders and files. But since enterprise deployment usually is done via machine-context (built-in user „system“ is used), other users won’t have write access to the folders or files which are created by „system“. Simply said, User2 can access User2-files but not User1-files.

Can we agree on that?

**Comment by droidmonkey** [comment]
Please take this debate elsewhere, thank you
